File: src/backlog.ts
import { mkdir, readdir, rm } from 'node:fs/promises'
import { z } from 'zod'
import type { RepoContext } from './config.ts'
import { type TaskDelivery, TaskDeliverySchema } from './delivery.ts'

// The repo's backlog: vetted intents waiting to be dispatched to a worktree.
// One JSON file per entry under <repo-state>/backlog/. This is what you feed; the
// dispatcher (next) drains it by spawning a worker worktree per entry.

const EntrySchema = z.object({
  id: z.string(),
  intent: z.string(),
  verify: z.string().nullable().default(null),
  createdAt: z.string(),
  chainId: z.string().nullable().default(null),
  chainOrder: z.number().int().nonnegative().nullable().default(null),
  previousId: z.string().nullable().default(null),
  delivery: TaskDeliverySchema.nullable().default(null),
})
export type BacklogEntry = z.infer<typeof EntrySchema>

export type AddBacklogOptions = {
  suggestedId?: string
  chainId?: string
  chainOrder?: number
  previousId?: string
  delivery?: TaskDelivery
}

function slugify(text: string): string {
  const slug = text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 40)
    .replace(/-+$/g, '')
  return slug.length > 0 ? slug : 'task'
}

async function entryFiles(dir: string): Promise<string[]> {
  try {
    return (await readdir(dir)).filter((f) => f.endsWith('.json')).sort()
  } catch {
    return []
  }
}

export async function addBacklog(
  ctx: RepoContext,
  intent: string,
  verify: string | null,
  options: AddBacklogOptions = {}
): Promise<BacklogEntry> {
  await mkdir(ctx.backlogDir, { recursive: true })
  // Descriptive, number-free id; disambiguate same-named entries with a -N suffix.
  const slug = slugify(options.suggestedId ?? intent.trim().split('\n', 1)[0] ?? '')
  const files = await entryFiles(ctx.backlogDir)
  let id = slug
  const existingPath = `${ctx.backlogDir}/${id}.json`
  if (files.includes(`${id}.json`) && options.chainId) {
    const existing = EntrySchema.safeParse(await Bun.file(existingPath).json())
    if (
      existing.success &&
      existing.data.chainId === options.chainId &&
      existing.data.chainOrder === (options.chainOrder ?? null)
    ) {
      return existing.data
    }
  }
  for (let n = 2; files.includes(`${id}.json`); n++) {
    id = `${slug}-${n}`
  }
  const entry: BacklogEntry = {
    id,
    intent: intent.trim(),
    verify,
    createdAt: new Date().toISOString(),
    chainId: options.chainId ?? null,
    chainOrder: options.chainOrder ?? null,
    previousId: options.previousId ?? null,
    delivery: options.delivery ?? null,
  }
  await Bun.write(`${ctx.backlogDir}/${id}.json`, `${JSON.stringify(entry, null, 2)}\n`)
  return entry
}

// Ordered oldest-first by createdAt (ids no longer carry a sortable number).
export async function loadBacklog(ctx: RepoContext): Promise<BacklogEntry[]> {
  const entries: BacklogEntry[] = []
  for (const file of await entryFiles(ctx.backlogDir)) {
    const parsed = EntrySchema.safeParse(await Bun.file(`${ctx.backlogDir}/${file}`).json())
    if (parsed.success) {
      entries.push(parsed.data)
    }
  }
  return entries.sort((a, b) => a.createdAt.localeCompare(b.createdAt))
}

export function orderedChainEntries(
  entries: BacklogEntry[],
  chainId: string | undefined
): BacklogEntry[] {
  if (!chainId) {
    return entries
  }
  return entries
    .filter((entry) => entry.chainId === chainId)
    .sort((a, b) => (a.chainOrder ?? 0) - (b.chainOrder ?? 0))
}

export async function removeBacklog(
  ctx: RepoContext,
  query: string
): Promise<{ removed: BacklogEntry } | { ambiguous: BacklogEntry[] } | null> {
  const entries = await loadBacklog(ctx)
  const exact = entries.find((e) => e.id === query)
  const matches = exact ? [exact] : entries.filter((e) => e.id.includes(query))
  const entry = matches[0]
  if (!entry) {
    return null
  }
  if (matches.length > 1) {
    return { ambiguous: matches }
  }
  await rm(`${ctx.backlogDir}/${entry.id}.json`, { force: true })
  return { removed: entry }
}

File: src/backlog.rs
use std::fs;
use std::io;
use std::path::Path;

use regex::Regex;
use serde::{Deserialize, Serialize};

use crate::clock::now_iso;
use crate::config::RepoContext;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct BacklogEntry {
    pub id: String,
    pub intent: String,
    #[serde(default)]
    pub verify: Option<String>,
    #[serde(rename = "createdAt")]
    pub created_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RemoveBacklogResult {
    Removed(BacklogEntry),
    Ambiguous(Vec<BacklogEntry>),
}

fn slugify(text: &str) -> String {
    let re = Regex::new(r"[^a-z0-9]+").expect("slug regex compiles");
    let lower = text.to_lowercase();
    let slug = re.replace_all(&lower, "-");
    let slug = slug
        .trim_matches('-')
        .chars()
        .take(40)
        .collect::<String>()
        .trim_end_matches('-')
        .to_string();
    if slug.is_empty() {
        "task".to_string()
    } else {
        slug
    }
}

fn entry_files(dir: &str) -> Vec<String> {
    let Ok(entries) = fs::read_dir(dir) else {
        return Vec::new();
    };
    let mut files: Vec<String> = entries
        .filter_map(Result::ok)
        .filter_map(|entry| entry.file_name().into_string().ok())
        .filter(|name| name.ends_with(".json"))
        .collect();
    files.sort();
    files
}

pub fn add_backlog(
    ctx: &RepoContext,
    intent: &str,
    verify: Option<String>,
) -> io::Result<BacklogEntry> {
    fs::create_dir_all(&ctx.backlog_dir)?;
    let slug = slugify(intent.trim().lines().next().unwrap_or_default());
    let files = entry_files(&ctx.backlog_dir);
    let mut id = slug.clone();
    for n in 2.. {
        if !files.contains(&format!("{id}.json")) {
            break;
        }
        id = format!("{slug}-{n}");
    }
    let entry = BacklogEntry {
        id: id.clone(),
        intent: intent.trim().to_string(),
        verify,
        created_at: now_iso(),
    };
    let json = serde_json::to_string_pretty(&entry).map_err(io::Error::other)?;
    fs::write(
        Path::new(&ctx.backlog_dir).join(format!("{id}.json")),
        format!("{json}\n"),
    )?;
    Ok(entry)
}

pub fn load_backlog(ctx: &RepoContext) -> io::Result<Vec<BacklogEntry>> {
    let mut entries = Vec::new();
    for file in entry_files(&ctx.backlog_dir) {
        let path = Path::new(&ctx.backlog_dir).join(file);
        let text = fs::read_to_string(path)?;
        if let Ok(entry) = serde_json::from_str::<BacklogEntry>(&text) {
            entries.push(entry);
        }
    }
    entries.sort_by(|a, b| a.created_at.cmp(&b.created_at));
    Ok(entries)
}

pub fn remove_backlog(ctx: &RepoContext, query: &str) -> io::Result<Option<RemoveBacklogResult>> {
    let entries = load_backlog(ctx)?;
    let exact = entries.iter().find(|entry| entry.id == query).cloned();
    let matches: Vec<BacklogEntry> = exact.map(|entry| vec![entry]).unwrap_or_else(|| {
        entries
            .into_iter()
            .filter(|entry| entry.id.contains(query))
            .collect()
    });
    let Some(entry) = matches.first() else {
        return Ok(None);
    };
    if matches.len() > 1 {
        return Ok(Some(RemoveBacklogResult::Ambiguous(matches)));
    }
    fs::remove_file(Path::new(&ctx.backlog_dir).join(format!("{}.json", entry.id))).ok();
    Ok(Some(RemoveBacklogResult::Removed(entry.clone())))
}
